### PR TITLE
[TS] LPS-127207

### PR DESCRIPTION
**Pull request by balazssk**
Hi @vendeltoreki ,

I've found that Changeset "recursion" in _collectChildrenStagedModels does not carry over the current stagedModel, instead it recalls the parentStagedModel, resulting in an endless cycle. This was instroduced in https://github.com/shuyangzhou/liferay-portal/pull/9044/commits/056fede5a6b436ff2c21a74ec7d8f474bbec3a97 where an original recursion was transformed into the queue solution.
I've also found that PublishFolderMVCActionCommand is also using recursion to track all descendants of the parent folder. I've removed this, as Changeset is already handling this case.

Can you please review?
https://issues.liferay.com/browse/LPS-127207

Thanks,
Balázs

### CODE REVIEW

**Comment by vendeltoreki** [comment]
ci:test:sf

**Comment by vendeltoreki** [comment]
ci:test:relevant

**Comment by vendeltoreki** [comment]
ci:test:staging

**Comment by vendeltoreki** [comment]
Hi @balazssk ,

Thanks for the fix. I think it's good to go as a hotfix.

I've found some other classes where there's some double recursion might be involved:
https://github.com/liferay/liferay-portal/blob/master/modules/apps/bookmarks/bookmarks-web/src/main/java/com/liferay/bookmarks/web/internal/portlet/action/PublishFolderMVCActionCommand.java
https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/portlet/action/PublishFolderMVCActionCommand.java

The test failures seem to be unrelated.

Regards,
Vendel

**Comment by vendeltoreki** [comment]
ci:test:relevant

**Comment by balazssk** [comment]
Hi @vendeltoreki ,
Thanks for checking. I've also found that bookmarks and DL are using a similar builder to determine the hierarchy of  staged models.
Bookmarks doesn't have any specific code for the hierarchy's supplier function and the LocalService call itself is not recursive.
Document Library  doesn't follow this. Instead of letting Changeset handle the recursion, it has its own code do it.

So to sum it up, DL was never affected by this issue. BM and JA was affected by the Changeset code going into infinite recursion. Finally, JA had an extra layer of recursion, which itself is only a logical flaw.

**Comment by vendeltoreki** [comment]
Thanks, @balazssk . I've tested the code and it works correctly. Forwarding the request.

**Comment by vendeltoreki** [comment]
ci:forward
